src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,89 @@ namespace cryptonote
+        const boost::filesystem::directory_iterator end_itr;
+        for (boost::filesystem::directory_iterator iter(power_supply_class_path); iter != end_itr; ++iter)
         {
-          power_supply_path = path;
-          break;
+          const boost::filesystem::path & power_supply_path = iter->path();
+          if (boost::filesystem::is_directory(iter->path()))
+          {
+            std::string power_supply_present;
+            epee::file_io_utils::load_file_to_string((power_supply_path / "/present").string(), power_supply_present);
+            if (!boost::starts_with(power_supply_present, "1")) {
+              LOG_PRINT_L4("Power supply not present at " << power_supply_path);
+              continue;
+            }
+
+            boost::filesystem::path power_supply_type_path = power_supply_path / "/type";
+            if (boost::filesystem::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type;
+              epee::file_io_utils::load_file_to_string(power_supply_type_path.string(), power_supply_type);
+
+              // If there is an AC adapter that's present and online we can break early
+              if (boost::starts_with(power_supply_type, "Mains"))
+              {
+                boost::filesystem::path power_supply_online_path = power_supply_path / "/online";
+                if (boost::filesystem::is_regular_file(power_supply_online_path))
+                {
+                  std::string power_supply_online;
+                  epee::file_io_utils::load_file_to_string(power_supply_online_path.string(), power_supply_online);
+
+                  if (boost::starts_with(power_supply_online, "1"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else
+                  {
+                    continue;
+                  }
+                }
+              }
+              else if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                boost::filesystem::path power_supply_status_path = power_supply_path / "/status";
+                if (boost::filesystem::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status;
+                  epee::file_io_utils::load_file_to_string(power_supply_status_path.string(), power_supply_status);
+
+                  // Possible status are Charging, Full, Discharging, Not Charging, and Unknown
+                  // We are only need to handle negative states right now
+
+                  if (on_battery &&
+                      is_charging == boost::logic::indeterminate &&
+                      (boost::starts_with(power_supply_status, "Charging") || boost::starts_with(power_supply_status, "Full")))
+                  {
+                    on_battery = boost::logic::tribool(false);
+                    is_charging = boost::logic::tribool(true);
```

Comment:
Status full does not mean it's charging from my understanding. It can be full for a short time with AC unplugged.
These extra variables are un-needed. Please see my comment describing the logic in pseudo-code as I think this whole function should be implemented like that as it handles all the cases.